theano/gof/fg.py

```diff
@@ -320,13 +321,15 @@ def __remove_clients__(self, r, clients_to_remove,
                               if output.clients or output in self.outputs]
             # If the apply node is not used and is not an output
             if not used_or_output:
-                self.apply_nodes.remove(apply_node)
```

Comment:
Why did you added this if? The if above on line 322 should be enough. If your if fail, it mean there is a bug problem and we want to crash, the graph isn't in a good condition.

The other change you did here seem good.